File: app/src/main/java/com/theappsolutions/boilerplate/customviews/RecyclerViewWithEmptyMode.java
package com.theappsolutions.boilerplate.customviews;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;

import com.annimon.stream.Optional;

/**
 * @author Dmytro Yakovlev d.yakovlev@theappsolutions.com
 * @copyright (c) 2017 TheAppSolutions. (https://theappsolutions.com)
 */
public class RecyclerViewWithEmptyMode extends RecyclerView {

    private Optional<OnUpdateListener> onUpdateListener = Optional.empty();

    @Nullable
    View emptyView;
    final AdapterDataObserver observer = new AdapterDataObserver() {
        @Override
        public void onChanged() {
            super.onChanged();
            checkIfEmpty();
            onUpdateListener.ifPresent(OnUpdateListener::onUpdate);
        }
    };

    public RecyclerViewWithEmptyMode(Context context) {
        super(context);
    }

    public RecyclerViewWithEmptyMode(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public RecyclerViewWithEmptyMode(Context context, AttributeSet attrs,
                                     int defStyle) {
        super(context, attrs, defStyle);
    }

    void checkIfEmpty() {
        if (emptyView != null) {
            emptyView.setVisibility(getAdapter() == null || getAdapter().getItemCount() > 0 ? GONE :
                    VISIBLE);
        }
    }

    @Override
    public void setAdapter(@Nullable Adapter adapter) {
        final Adapter oldAdapter = getAdapter();
        if (oldAdapter != null) {
            oldAdapter.unregisterAdapterDataObserver(observer);
        }
        super.setAdapter(adapter);
        if (adapter != null) {
            adapter.registerAdapterDataObserver(observer);
        }
    }

    public void setEmptyView(@Nullable View emptyView) {
        this.emptyView = emptyView;
        checkIfEmpty();
    }

    public interface OnUpdateListener {
        void onUpdate();
    }

    public void setOnUpdateListener(OnUpdateListener onUpdateListener) {
        this.onUpdateListener = Optional.of(onUpdateListener);
    }
}

File: app/src/main/java/com/theappsolutions/boilerplate/customviews/RecyclerViewWithEmptyMode.kt
package com.theappsolutions.boilerplate.customviews

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View

import com.annimon.stream.Optional

/**
 * @author Dmytro Yakovlev d.yakovlev@theappsolutions.com
 * @copyright (c) 2017 TheAppSolutions. (https://theappsolutions.com)
 */
class RecyclerViewWithEmptyMode : RecyclerView {

    private var onUpdateListener: OnUpdateListener? = null
    private lateinit var emptyView: View

    internal fun getObserver() = object : RecyclerView.AdapterDataObserver() {
        override fun onChanged() {
            super.onChanged()
            checkIfEmpty()
            onUpdateListener?.onUpdate()
        }
    }

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {}

    private fun checkIfEmpty() {
        if (emptyView != null) {
            emptyView.visibility = if (adapter == null || adapter.itemCount > 0) {
                View.GONE
            } else {
                View.VISIBLE
            }
        }
    }

    override fun setAdapter(adapter: RecyclerView.Adapter<*>?) {
        val oldAdapter = getAdapter()
        oldAdapter?.unregisterAdapterDataObserver(getObserver())
        super.setAdapter(adapter)
        adapter?.registerAdapterDataObserver(getObserver())
    }

    fun setEmptyView(emptyView: View) {
        this.emptyView = emptyView
        checkIfEmpty()
    }

    interface OnUpdateListener {
        fun onUpdate()
    }

    fun setOnUpdateListener(onUpdateListener: OnUpdateListener) {
        this.onUpdateListener = onUpdateListener
    }
}
